library/spdm_responder_lib/libspdm_rsp_set_certificate.c

```diff
@@ -0,0 +1,98 @@
+ *                                     On input, it means the size in bytes of response data buffer.
+ *                                     On output, it means the size in bytes of copied response data buffer if RETURN_SUCCESS is returned,
+ *                                     and means the size in bytes of desired response data buffer if RETURN_BUFFER_TOO_SMALL is returned.
+ * @param  response                     A pointer to the response data.
+ *
+ * @retval RETURN_SUCCESS               The request is processed and the response is returned.
+ * @retval RETURN_BUFFER_TOO_SMALL      The buffer is too small to hold the data.
+ * @retval RETURN_DEVICE_ERROR          A device error occurs when communicates with the device.
+ * @retval RETURN_SECURITY_VIOLATION    Any verification fails.
+ **/
+libspdm_return_t libspdm_get_response_set_certificate(void *context, size_t request_size,
+                                                      const void *request,
+                                                      size_t *response_size,
+                                                      void *response)
+{
+    spdm_set_certificate_request_t *spdm_request;
+    spdm_set_certificate_response_t *spdm_response;
+
+    libspdm_context_t *spdm_context;
+    bool result;
+    uint8_t slot_id;
+    bool need_reset;
+
+    spdm_context = context;
+    spdm_request = (void*)request;
+
+    if (spdm_request->header.spdm_version != libspdm_get_connection_version(spdm_context)) {
+        return libspdm_generate_error_response(spdm_context,
+                                               SPDM_ERROR_CODE_VERSION_MISMATCH, 0,
+                                               response_size, response);
+    }
+
+    if (spdm_context->response_state != LIBSPDM_RESPONSE_STATE_NORMAL) {
+        return libspdm_responder_handle_response_state(spdm_context,
+                                                       spdm_request->header.request_response_code,
+                                                       response_size, response);
+    }
+
+    if (request_size <= sizeof(spdm_set_certificate_request_t)) {
+        return libspdm_generate_error_response(spdm_context,
+                                               SPDM_ERROR_CODE_INVALID_REQUEST, 0,
+                                               response_size, response);
+    }
+
+    slot_id = spdm_request->header.param1 & SPDM_GET_CERTIFICATE_REQUEST_SLOT_ID_MASK;
+    if (slot_id >= spdm_context->local_context.slot_count) {
+        return libspdm_generate_error_response(spdm_context,
+                                               SPDM_ERROR_CODE_INVALID_REQUEST, 0,
+                                               response_size, response);
+    }
+
+    /* set certificate to NV*/
+    result = libspdm_set_certificate_to_nv(slot_id, spdm_request + 1,
+                                           request_size - sizeof(spdm_set_certificate_request_t),

```

Comment:
This code is not right. We need parse cert chain to get full size.

Reply:
Yes. I parse the cert chain in l`ibspdm_set_certificate_to_nv` to get the actual total size.